message/messaging_service.cc

```diff
@@ -677,15 +678,24 @@ messaging_service::get_rpc_client_idx(messaging_verb verb) const {
         return idx;
     }
 
-    // A statement or statement-ack verb
-    const auto curr_sched_group = current_scheduling_group();
-    for (unsigned i = 0; i < _connection_index_for_tenant.size(); ++i) {
-        if (_connection_index_for_tenant[i].sched_group == curr_sched_group) {
-            // i == 0: the default tenant maps to the default client indexes belonging to the interval
-            // [PER_SHARD_CONNECTION_COUNT, PER_SHARD_CONNECTION_COUNT + PER_TENANT_CONNECTION_COUNT).
-            idx += i * PER_TENANT_CONNECTION_COUNT;
-            break;
+    std::function<std::optional<uint8_t>(const scheduling_group)> find_stmt_tenant_idx = [this, idx, &find_stmt_tenant_idx] (const scheduling_group curr_sched_group) -> std::optional<uint8_t> {
+        for (unsigned i = 0; i < _connection_index_for_tenant.size(); ++i) {
+            if (_connection_index_for_tenant[i].sched_group == curr_sched_group) {
+                if (!_connection_index_for_tenant[i].enabled && curr_sched_group != default_scheduling_group()) {
+                    return find_stmt_tenant_idx(default_scheduling_group());
+                }
+
+                // i == 0: the default tenant maps to the default client indexes belonging to the interval
+                // [PER_SHARD_CONNECTION_COUNT, PER_SHARD_CONNECTION_COUNT + PER_TENANT_CONNECTION_COUNT).
+                return {idx + i * PER_TENANT_CONNECTION_COUNT};
+            }
         }
+        return std::nullopt;
+    };
+ 
+    // A statement or statement-ack verb
+    if (auto stmt_tenant_idx = find_stmt_tenant_idx(current_scheduling_group())) {
```

Comment:
Why wrap into a function? Wouldn't the old fall-back of `return idx` work just as fine?

Reply:
It still works the same but the tenant is disabled, then the function is run once again looking for default tenant. That's I've extracted it to a function.

Reply:
I don't think it will work the same. What I think that Botond imagined is just changing

```c++
if (_connection_index_for_tenant[i].sched_group == curr_sched_group) {
```

to

```c++
if (_connection_index_for_tenant[i].sched_group == curr_sched_group && _connection_index_for_tenant[i].enabled) {
```

and letting `return idx` handle it if the tenant is not enabled. If the maintenance tenant is not enabled, the `return idx` will return a value from range `[0, PER_TENANT_CONNECTION_COUNT)` - or, in other terms, the one that would correspond to the tenant with idx `0` (`"$user"`).

In the current solution, we will look for a connection corresponding to the tenant assigned to the default scheduling group (`"$system"`). This tenant has index **1**, not 0, so it will return something from range `[PER_TENANT_CONNECTION_COUNT, 2*PER_TENANT_CONNECTION_COUNT)`.

@denesb which tenant is preferred in this case, `$user` or `$system`?

Reply:
Removed the wrapper function. If no tenant is found, then `$user` is used.

When this PR will be merged to enterprise, this code has to be changed to use default tenant when tenant for current scheduling group is disabled and to not fall into [service levels tenant](https://github.com/scylladb/scylla-enterprise/blob/enterprise/message/messaging_service.cc#L736-L761).

Reply:
> @denesb which tenant is preferred in this case, $user or $system?

We have a "default" tenant, the first tenant in the tenant configuration. That one should be used, to allow the default tenant to be selected by the configuration, instead of hard-coding it.